### ISSUE TITLE
No module named 'utils' or Import "utils" could not be resolved

### ISSUE THREAD

**Issue by ShannonTown** [tpecot/DeepLearningForBiologists_FromClassificationToSegmentation#2]
Hi, I'm having problem running the Training2D.ipynb or Training.ipynb, either it says no module named 'utils' (in jupyter notebook, **1st scenario**), or the first cell can run and shows a warning "import 'utils" could not be resolved, and a warning about tensorflow "The name tf.losses.sparse_softmax_cross_entropy is deprecated. Please use tf.compat.v1.losses.sparse_softmax_cross_entropy instead." (in VS Code **2nd scenario**)

After running the first cell, I ran the second cell for setting parameters, it shows "name 'training_parameters_stardist_interface_2D' is not defined" (1st scenario) or show pending forever (2nd scenario). What do you think might have caused this issue? Thank you!

**Comment by ShannonTown** [comment]
PS, I'm using a virtual conda environment in which tensorflow, torch, etc is installed

**Comment by tpecot** [comment]
Hi,

It looks like the code cannot access utils.py. Is utils.py defined in the folder where you hace Training2D.ipynb?

**Comment by ShannonTown** [comment]
Indeed in scenario 1 when I used jupyter notebook utils.py is not in the folder, when I include this file it's working! In scenario 2 it still doesn't work in VS code, the parameter setting interface would not show up. I will keep using jupyter notebook now. 

Thank you so much for the very helpful repositories!

**Comment by tpecot** [comment]
Great, enjoy!

**Comment by ShannonTown** [comment]
Hi, I created a new virtual environment fixing some package versions and a missing dll file, now VS code works too. The UNet training script works fine, however, Stardist training script gives the following error no matter with my own data or your example data, not matter using Jupyter notebook or VS code. It can start training but stops at the second line (mostly if not always). Would you please help troubleshoot? Is this due to limited memory of my GPU?

**Comment by ShannonTown** [comment]
![image](https://github.com/tpecot/DeepLearningForBiologists_FromClassificationToSegmentation/assets/30399381/5967068e-7856-4f34-a3e5-9e1a281bbbeb)
![image](https://github.com/tpecot/DeepLearningForBiologists_FromClassificationToSegmentation/assets/30399381/34fb24ae-a994-4b0e-b959-c5dadb4ca5e4)

**Comment by ShannonTown** [comment]
I'm new to ML, and not sure if I should choose UNet or Stardist for my data. I've tried UNet with 12 training images but the trained model wasn't good, I would like to test Stardist before using more data to train with an approach that seems better for my data. Thanks very much for your help!

**Comment by tpecot** [comment]
Hi, it seems that the first epoch goes through but it fails after it, probably the images in the validation dataset do not have the correct dimension given the error message. 
Choosing Stardist or U-Net depends on what you want to do. If you want to segment individual objects that have a convex shape, then Stardist is a good option. If you want to get a class at each pixel, then Stardist is not useful and U-Net will fit perfectly. Npw, even if you want to segment idividual objects, U-Net might give you a good solution with a proper post-processing. You can for example add a class for object contours that will help separate touching objects. Hope this will help!

**Comment by ShannonTown** [comment]
Hi, I really appreciate the elaboration. For Stardist, I think my validation dataset is correct since it works for UNet, and the script gives the same error with your example polyp_nuclei data. Nonetheless, I trained with UNet using more data, and got a model that works most of the time, for the images where it doesn't work so well, I'm assuming it can be improved with some post-processing you mentioned. Would you mind pointing me to a resource for post processing if you happen to know it?

**Comment by ShannonTown** [comment]
PS, the training and validation masks I provided have nuclei as white (value 0), and background as black (value 255), but somehow the UNet model generated masks with inverted values.

**Comment by tpecot** [comment]
Hi! As I said, U-Net and Stardist do not do the same thing, so masks are different. With U-Net, you estimate a class at each pixel, in your case you have 2 classes (background and nuclei), with Stardist you idenntify individual objects so each object is defined by a different intensity. Using the U-Net masks for Stardist will not make sense as you'll only have one nuclus that is split many times. If you want to see how to use these codes, have a look at the Youtube videos I made about them, I also show the kind of masks that are used with U-Net for segmenting tissues, nuclei (adding a class for nuclei contours) and for Stardist.

**Comment by ShannonTown** [comment]
I see, so UNet would be the better option for my application at the moment. I have watched a bunch of youtube videos and tried many things that didn't work, until I found your videos. Apologies if I missed anything in your videos and thanks again for all the help.

**Comment by tpecot** [comment]
Glad to see it helped!
